### ISSUE TITLE
CNN predicting only one label regardless of the input

### ISSUE THREAD

**Issue by shashankgwl** [microsoft/CNTK#3730]
I'm tring to predict three labels
`1, 2, 3`
I've 1000 images for each label, each of the size 50x50, which I've carefully crafted using openCV, finally during prediction, I use the same open openCV code to capture a new image from camera, pre-process it using the same code, and , feed to `eval` method.

**Entire zipped data is available to download from** [here](https://github.com/shashankgwl/Cntk.Gesture.Recognition/blob/master/gestureData.7z).

Please note, there are very few files for label '3', because only this label was getting predicted earlier for any input, hence I reduced the training data for this label. however, nothing changed, it started predicting lable '2' for every input.

**However, no matter what input, the output is only one class (usually the last one).**

> Model : the model definition looks like below

```
def create_basic_model(input, out_dims):
    with C.layers.default_options(init=C.glorot_uniform(), activation=C.relu):
        model = C.layers.Sequential([
            C.layers.For(range(3), lambda i: [
                C.layers.Convolution((5,5), [32,32,64][i], pad=True),
                C.layers.MaxPooling((3,3), strides=(2,2))
                ]),
            C.layers.Dense(64),
            C.layers.Dropout(0.25),
            C.layers.Dense(out_dims, activation=C.softmax)
        ])

        return model(input)
```

Entire definition is actually copied from CNTK 201B article.

> Reader definition : Below is how I'm reading all the files for training, t**here's no mean file, does this impact? I believe, I'm already doing a lot of pre-processing using OpenCV.**

```
def create_reader(map_file, train):
    #print("Reading map file:", map_file)
    
    
    # transformation pipeline for the features has jitter/crop only when training
    transforms = []
    # train uses data augmentation (translation only)
    if train:
        transforms += [
            xforms.crop(crop_type='randomside', side_ratio=0.8) 
        ]
    transforms += [
        xforms.scale(width=image_width, height=image_height, channels=num_channels, interpolations='linear')
    ]
    # deserializer
    return C.io.MinibatchSource(C.io.ImageDeserializer(map_file, C.io.StreamDefs(
        features = C.io.StreamDef(field='image', transforms=transforms), # first column in map file is referred to as 'image'
        labels   = C.io.StreamDef(field='label', shape=num_classes)      # and second as 'label'
    )))

```

> Training : Here's how the model is being trained.

```
input_var = C.input_variable((num_channels, image_height, image_width))
    label_var = C.input_variable((num_classes))

    # Normalize the input
    feature_scale = 1./ 255.
    input_var_norm = C.element_times(feature_scale, input_var)
    
    # apply model to input
    z = model_func(input_var_norm, out_dims=3)
    ce = C.cross_entropy_with_softmax(z, label_var)
    pe = C.classification_error(z, label_var)

    # training config
    epoch_size     = 50000
    minibatch_size = 20

    # Set training parameters
    lr_per_minibatch       = C.learning_rate_schedule([0.001]*10 + [0.003]*10 + [0.001], 
                                                      C.UnitType.minibatch, epoch_size)
    momentum_time_constant = C.momentum_as_time_constant_schedule(-minibatch_size/np.log(0.9))
    l2_reg_weight          = 0.001
    
    # trainer object
    #learner = C.sgd(z.parameters, lr = 0.01, minibatch_size = C.learners.IGNORE)
    learner = C.momentum_sgd(z.parameters, lr = lr_per_minibatch, momentum = momentum_time_constant, l2_regularization_weight=l2_reg_weight)
    progress_printer = C.logging.ProgressPrinter(tag='Training', num_epochs=max_epochs)
    trainer = C.Trainer(z, (ce, pe), [learner], [progress_printer])
```

> Learning performance 

It seems to be learning something, as you can see from the image below.
![image](https://user-images.githubusercontent.com/2750603/62000425-a15f1900-b0a4-11e9-8239-e1b9c1a3c45b.png)

> Evaluation : Finally a new imaged is predicted using below code.

```
while 1:    
    ret, frame = cap.read()
    frame = cv.flip(frame, 1)
    hsv = cv.cvtColor(frame, cv.COLOR_BGR2HSV)
    mask1 = cv.inRange(hsv,np.array([0,30,60]),np.array([20,150,255]))
    newmasked = cv.bitwise_and(frame,frame,mask=mask1)
    newmasked= cv.cvtColor(newmasked,cv.COLOR_BGR2GRAY)
    #print(newmasked.shape)
    blurred = cv.GaussianBlur(newmasked,(5,5),0)
    
    canny = cv.Canny(blurred,0,255)
    kernel = np.ones((2,2),np.uint8)
    opening = cv.morphologyEx(canny, cv.MORPH_CLOSE, kernel)
    
    #cv.rectangle(opening,(350,10),(610,228),(255,255,255),1)
    
    
    try:
        ROI = opening[10:228, 350:600]
        image2predict = cv.resize(ROI,(50,50))
        eval(pred,np.array(image2predict,dtype=np.float32))
```

> Eval method
```
def eval(pred_op, image_data):
    #print("shape of image is {0}".format(image_data.shape))
    label_lookup = ["One", "two", "three"]
    image_data = np.ascontiguousarray(image_data.reshape(1,50,50),dtype=np.float32)
    
    result = np.squeeze(pred_op.eval({pred_op.arguments[0]:[image_data]}))
    
    # Return top 3 results:
    top_count = 3
    
    result_indices = (-np.array(result)).argsort()[:top_count]
    res = result[result_indices[0]] * 100
    if(res > 90.):
        print("Prediction: {:10s}, confidence: {:.2f}%\n".format(label_lookup[result_indices[0]], result[result_indices[0]] * 100))
```

**Following things have already been tried**

1. Using epochs ranging from 5-30 (no impact).
2. using softmax and relu as activation in final Dense layer.
3. Generating more data for the labels which weren't getting predicted.
4. Tried learning rate 0.01 and 0.001
5. Mini-batch size , tried 64 (copied from article) and 20 (no impact).

**Comment by delzac** [comment]
If you are using `cross_entropy_with_softmax` loss then your last activation should be `None` and not `C.softmax`.

You can use some test data for label three to check if it outputs label three in actuality.

From your training loss graph, it seems like the training isn't completed yet.

The mean file doesn't matter.

**Comment by shashankgwl** [comment]
I've already tried that as part of original code copied from CNTK 201B. still not clue, the pattern I'm seeing is _last one wins_.
